<a href="https://colab.research.google.com/github/emreozan/ML-AI-Portfolio/blob/main/MNIST%3E97_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The goal of this exercise is to have a classifier that achieves over 97 percent accuracy  on the test set. 
Source: Hands on ML by Geron, Chapter 3, Exercise 1

# LIBRARIES

In [1]:
#Libraries used
from sklearn.datasets import fetch_openml
from sklearn.neighbors import KNeighborsClassifier
from  sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

import matplotlib as mpl
import matplotlib.pyplot as plt

from scipy.ndimage.interpolation import shift

# Data Loading

In [2]:
#load the dataset
mnist=fetch_openml("mnist_784",version=1)

In [3]:
X,y =mnist["data"],mnist["target"]

In [4]:
X.shape, y.shape

((70000, 784), (70000,))

# Without data augmentation

## 1000 data points

In [ ]:
#X_train, X_test, y_train, y_test=X[:60000],X[60000:],y[:60000],y[60000:]

X_train, X_test, y_train, y_test=X[:1000],X[1000:1100],y[:1000],y[1000:1100]


Obviously we are not there yet→ Hyperparameter Optimization through grid search

In [ ]:
parameters={
    "weights":("uniform","distance"),"n_neighbors":[1,10]
}

In [ ]:
k_class=KNeighborsClassifier()
k_class_grid=GridSearchCV(k_class,parameters)
k_class_grid.fit(X_train,y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='deprecated', n_jobs=None,
             param_grid={'n_neighbors': [1, 10],
                         'weights': ('uniform', 'distance')},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [ ]:
k_class_grid.best_params_

{'n_neighbors': 1, 'weights': 'uniform'}

In [ ]:
k_class_grid.best_score_


0.8619999999999999

0.86 was the result

## 10000 data points


In [ ]:
mnist=fetch_openml("mnist_784",version=1)
X,y =mnist["data"],mnist["target"]
X_train, X_test, y_train, y_test=X[:10000],X[10000:11000],y[:10000],y[10000:11000]

In [ ]:
parameters={
    "weights":("uniform","distance"),"n_neighbors":[3,4,5]
}
k_class=KNeighborsClassifier()
k_class_grid=GridSearchCV(k_class,parameters,cv=3,verbose=3)
k_class_grid.fit(X_train,y_train)

Fitting 3 folds for each of 6 candidates, totalling 18 fits
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ...... n_neighbors=3, weights=uniform, score=0.939, total=  31.4s
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   31.4s remaining:    0.0s


[CV] ...... n_neighbors=3, weights=uniform, score=0.942, total=  31.4s
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  1.0min remaining:    0.0s


[CV] ...... n_neighbors=3, weights=uniform, score=0.925, total=  31.8s
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.943, total=  31.1s
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.945, total=  31.4s
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.928, total=  31.3s
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.938, total=  31.9s
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.945, total=  31.7s
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.921, total=  31.0s
[CV] n_neighbors=4, weights=distance .................................
[CV] .

[Parallel(n_jobs=1)]: Done  18 out of  18 | elapsed:  9.3min finished


GridSearchCV(cv=3, error_score=nan,
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='deprecated', n_jobs=None,
             param_grid={'n_neighbors': [3, 4, 5],
                         'weights': ('uniform', 'distance')},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [ ]:
k_class_grid.best_params_

{'n_neighbors': 4, 'weights': 'distance'}

In [ ]:
k_class_grid.best_score_


0.9397994088551026

With 10000 data points, score increased to 0.93

## All data points 

In [ ]:
mnist=fetch_openml("mnist_784",version=1)
X,y =mnist["data"],mnist["target"]
X_train, X_test, y_train, y_test=X[:60000],X[60000:],y[:60000],y[60000:]

In [ ]:
parameters={
    "weights":("uniform","distance"),"n_neighbors":[3,4,5]
}
k_class=KNeighborsClassifier()
k_class_grid=GridSearchCV(k_class,parameters,cv=3,verbose=3)
k_class_grid.fit(X_train,y_train)

Fitting 3 folds for each of 6 candidates, totalling 18 fits
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ...... n_neighbors=3, weights=uniform, score=0.969, total=20.0min
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed: 20.0min remaining:    0.0s


[CV] ...... n_neighbors=3, weights=uniform, score=0.968, total=19.8min
[CV] n_neighbors=3, weights=uniform ..................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed: 39.8min remaining:    0.0s


[CV] ...... n_neighbors=3, weights=uniform, score=0.968, total=19.8min
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.970, total=19.8min
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.969, total=19.8min
[CV] n_neighbors=3, weights=distance .................................
[CV] ..... n_neighbors=3, weights=distance, score=0.969, total=19.8min
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.966, total=20.0min
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.966, total=19.8min
[CV] n_neighbors=4, weights=uniform ..................................
[CV] ...... n_neighbors=4, weights=uniform, score=0.967, total=19.8min
[CV] n_neighbors=4, weights=distance .................................
[CV] .

[Parallel(n_jobs=1)]: Done  18 out of  18 | elapsed: 358.5min finished


GridSearchCV(cv=3, error_score=nan,
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='deprecated', n_jobs=None,
             param_grid={'n_neighbors': [3, 4, 5],
                         'weights': ('uniform', 'distance')},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [ ]:
k_class_grid.best_params_

{'n_neighbors': 4, 'weights': 'distance'}

In [ ]:
k_class_grid.best_score_

0.9703500000000002

When all data points were used, score was slightly above 0.97